### ISSUE TITLE
Diff search includes files in search result totals

### ISSUE THREAD

**Issue by Galindo-J** [sourcegraph/sourcegraph#48507]
- **Sourcegraph version:** v4.4.2 and sourcegraph.com
- **Platform information:** Self hosted customer (4.4.2) & dotcom

#### Steps to reproduce:

1. Create a diff search targeting a specific file and string. 

[Example](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+file:%5E%5C.bazelrc+type:diff+ENABLE_BAZEL&patternType=standard&sm=1&groupBy=repo): `repo:^github\.com/sourcegraph/sourcegraph$ file:^\.bazelrc type:diff ENABLE_BAZEL`

#### Expected behavior:
The query should return 1 result as the string only occurs once in the file/diff. 

#### Actual behavior:

The query results include the content match and the files, resulting in 3 results. 

<img width="1142" alt="Screenshot 2023-03-01 at 4 35 19 PM" src="https://user-images.githubusercontent.com/54120980/222281467-df5f1256-36dd-43d3-a932-25c9f8d2cb8d.png">

If you would like immediate help on this, please email support@sourcegraph.com (you can still create the issue, but there are no [SLAs](https://about.sourcegraph.com/support/) on issues like there are for support requests).

**Comment by keegancsmith** [comment]
The way it works in zoekt is if we have content matches, we throw away the path matches. I was experimenting with changing that behaviour not so long ago for the purposes of ranking, but didn't realise this implication.

It feels like as we evolve how this is reported this is the correct behaviour. And then if a user only was interested in content or path, they could use select? WDYT?

**Comment by camdencheek** [comment]
Result counting is tricky. I'm not sure what I think 😄 

Starting with a few statements I _think_ are true:
- Where possible, we should highlight every matched field in a result. e.g. `repo:a file:b c` should highlight `a`, `b`, and `c`. For that specific search, we currently highlight `b` and `c`, but not `a`, which I consider a UX bug.
- There are some matchable properties that are not highlightable because they are not displayed in the UI. e.g. `repo:has.commit.after(yesterday)`. In this case, the logical "number of results" is definitely not equivalent to the number of highlights.
- For some result types (repo, symbol, path), we count the number of unique objects (repos, symbols, paths). For other result types (file, commit, diff), we count the number of "matches" within that type. I think this is inconsistent and somewhat central to the murkiness here. This is difficult because, for the most part, I expect people care more about the number of matches than the number of unique files that contain those matches.
- A "match" can be against any number of logical "fields" of a result type. You can match against the file contents, the file path, the file's repo, the file's owner, etc. There may be multiple matches against a single field.

Some motivating examples. How should we count the results for each of these?
- `type:repo sourcegraph cody` returns a repo (`github.com/sourcegraph/cody`) with a single matched field (name) and two matched ranges (`sourcegraph` and `cody`)
- `type:file repo:sourcegraph/sourcegraph file:search_results.go resolver` returns a file result with three matched fields (repo name, file name, content) and many matches with the content field
- `type:symbol repo:sourcegraph/sourcegraph file:search_results.go resolver` returns a handful of symbols grouped into a single file result. There is a match on repo name, a match on file name, and a match against each of the symbol names.
- `type:file graphqlbackend search_results.go` (assuming a merged `type:file` and `type:path`) returns a result with no content matches, but with a path match.

We may be able to rely on the new filters panel to answer some of the hierarchy questions. e.g. answer the "X matches in Y files in Z repos" thing. I'll think on that a bit

**Comment by keegancsmith** [comment]
Indeed! I think we can discuss this in person at merge (atleast platform team can). Just to link some extra likely outdated content:
- limitations in limiting (related to counting) https://github.com/sourcegraph/sourcegraph/issues/18298#issue-808779075
- investigation on how counts worked at a point in time https://github.com/sourcegraph/sourcegraph/issues/17573#issuecomment-767588099
